### PR TITLE
Test freq orth pilot

### PR DESCRIPTION
**Pull request by di-mu**

### CODE REVIEW

**Review by rdoosty — APPROVED** [review]

File: src/client/txrx/workers/txrx_worker_client_sim.cc [hunk 1]
```diff
@@ -30,7 +30,7 @@ TxRxWorkerClientSim::TxRxWorkerClientSim(
                  event_notify_q, tx_pending_q, tx_producer, notify_producer,
                  rx_memory, tx_memory, sync_mutex, sync_cond, can_proceed),
       tx_pkt_zeros_(config->PacketLength(), 0u),
-      tx_pkt_pilot_(config->PacketLength(), 0u) {
+      tx_pkt_pilot_(config->UeAntNum(), tx_pkt_zeros_) {
```
Comment (asedlmayr) [inline src/client/txrx/workers/txrx_worker_client_sim.cc:33]: Initialize this, independent of tx_pkt_zeros_.

File: src/client/txrx/workers/txrx_worker_client_sim.h [hunk 2]
```diff
@@ -40,6 +40,6 @@ class TxRxWorkerClientSim : public TxRxWorker {
 
   //Helper tx vectors
   std::vector<uint8_t> tx_pkt_zeros_;
-  std::vector<uint8_t> tx_pkt_pilot_;
+  std::vector<std::vector<uint8_t>> tx_pkt_pilot_;
```
Comment (asedlmayr) [inline src/client/txrx/workers/txrx_worker_client_sim.h:43]: see comments above.  Might want to add another dimension

File: src/common/config.cc [hunk 3]
```diff
@@ -1420,22 +1413,55 @@ void Config::GenData() {
                       pre_uint32.end());
   this->pilot_.resize(this->samps_per_symbol_);
 
-  if (kDebugPrintPilot == true) {
-    std::cout << "Pilot data: " << std::endl;
-    for (size_t i = 0; i < this->ofdm_data_num_; i++) {
-      std::cout << this->pilots_[i].re << "+1i*" << this->pilots_[i].im << ",";
+  this->pilot_ue_sc_.resize(ue_ant_num_);
```
Comment (asedlmayr) [inline src/common/config.cc:1416]: You can initialize these to the fixed size.

File: src/common/config.h [hunk 4]
```diff
@@ -378,6 +379,12 @@ class Config {
   inline std::vector<std::complex<int16_t>>& PilotCi16() {
     return this->pilot_ci16_;
   };
+  inline std::vector<std::complex<int16_t>>& PilotUeCi16(size_t ue_id) {
+    return this->pilot_ue_ci16_.at(ue_id);
+  };
+  inline const arma::uvec& PilotUeSc(size_t ue_id) const {
```
Comment (asedlmayr) [inline src/common/config.h:385]: this type necessary?

**Review by asedlmayr — COMMENTED** [review]
Looks good, added a couple suggestions

File: src/agora/dobeamweights.cc [hunk 5]
```diff
@@ -85,17 +85,77 @@ DoBeamWeights::~DoBeamWeights() {
   std::free(calib_gather_buffer_);
 }
 
-EventData DoBeamWeights::Launch(size_t tag) {
-  ComputeFullCsiBeams(tag);
+EventData DoBeamWeights::Launch(size_t tag) {  // Compute Full-CSI Beams
+  const size_t frame_id = gen_tag_t(tag).frame_id_;
```
Comment (asedlmayr) [inline src/agora/dobeamweights.cc:89]: I would put the function call back for now.    Harder to track changes inside the function

File: src/agora/dobeamweights.cc [hunk 6]
```diff
@@ -85,17 +85,77 @@ DoBeamWeights::~DoBeamWeights() {
...
+  const size_t base_sc_id = gen_tag_t(tag).sc_id_;
+  const size_t frame_slot = frame_id % kFrameWnd;
+  if (kDebugPrintInTask) {
+    std::printf("In doZF thread %d: frame: %zu, base subcarrier: %zu\n", tid_,
+                frame_id, base_sc_id);
+  }
+  const size_t num_subcarriers =
+      cfg_->FreqOrthogonalPilot()
```
Comment (asedlmayr) [inline src/agora/dobeamweights.cc:97]: Why is this 1 for cfg_->FreqOrthogonalPilot() ?  Why 1?  Add name or clean up

File: src/agora/dobeamweights.h [hunk 7]
```diff
@@ -51,16 +51,18 @@ class DoBeamWeights : public Doer {
   EventData Launch(size_t tag) override;
 
  private:
-  void ComputePartialCsiBeams(size_t tag);
+  static inline void PartialTransposeGather(size_t cur_sc_id, float* src,
```
Comment (asedlmayr) [inline src/agora/dobeamweights.h:54]: Not sure why these are static?  probably doesn't hurt but I would hid them in the .cc file as static (not part of the class) or you can make them private member functions (non static).

File: src/agora/dobeamweights.cc [hunk 8]
```diff
@@ -376,8 +440,9 @@ static inline void PartialTransposeGather(size_t cur_sc_id, float* src,
 
 // Gather data of one symbol from partially-transposed buffer
 // produced by dofft
-static inline void TransposeGather(size_t cur_sc_id, float* src, float*& dst,
-                                   size_t bs_ant_num, size_t ofdm_data_num) {
+inline void DoBeamWeights::TransposeGather(size_t cur_sc_id, float* src,
```
Comment (asedlmayr) [inline src/agora/dobeamweights.cc:443]: Not sure if a &*  is really necessary.  Check the inputs and outputs and see if you can simplify the types

File: src/agora/dobeamweights.cc [hunk 9]
```diff
@@ -206,11 +266,14 @@ float DoBeamWeights::ComputePrecoder(size_t frame_id, size_t cur_sc_id,
     }
   }
   mat_ul_beam = mat_ul_beam_tmp;
+  if (kEnableMatLog) {
+    phy_stats_->UpdateUlBeam(frame_id, cur_sc_id, mat_ul_beam.st());
+  }
   float rcond = -1;
```
Comment (asedlmayr) [inline src/agora/dobeamweights.cc:272]: Why?  not necessary now

File: src/common/phy_stats.cc [hunk 10]
```diff
@@ -9,6 +9,9 @@
 
 #include "logger.h"
 
+//set the number of recorded subcarrier data samples
+static constexpr size_t kNumRecSc = 4;
```
Comment (asedlmayr) [inline src/common/phy_stats.cc:13]: Why is this 4?  why would you set it to anything else.  does this make sense to be hard coded?

File: src/common/phy_stats.cc [hunk 11]
```diff
@@ -381,18 +395,20 @@ void PhyStats::RecordDlPilotSnr(size_t frame_id) {
   }
 }
 
-void PhyStats::RecordDlCsi(size_t frame_id, size_t num_rec_sc,
+void PhyStats::RecordDlCsi(size_t frame_id,
```
Comment (asedlmayr) [inline src/common/phy_stats.cc:398]: This seems to make more sense as an input param to the function.

**Review by asedlmayr — COMMENTED** [review]
Added a few new comments

File: src/agora/dofft.cc [hunk 12]
```diff
@@ -198,14 +198,40 @@ EventData DoFFT::Launch(size_t tag) {
...
                                    fft_inout_);
+      }
     }
-    const size_t ue_id = pilot_symbol_id;
-    PartialTranspose(csi_buffers_[frame_slot][ue_id], ant_id,
+    PartialTranspose(csi_buffers_[frame_slot][pilot_symbol_id], ant_id,
                      SymbolType::kPilot);
+
+    // Expand partial CSI to full CSI per UE
+    if (cfg_->FreqOrthogonalPilot() && pilot_symbol_id == 0) {
```
Comment (asedlmayr) [inline src/agora/dofft.cc:216]: Change to pilot_symbol_id = last.    Not 0

**Comment by asedlmayr** [comment]
Branch got polluted, switching pr to older commit

**Pull request closed without merging**